File: TopCoder/python/arr/SortStackUsingTemporaryStack.py
#http://www.geeksforgeeks.org/sort-stack-using-temporary-stack/

def sort(stack):
    tempStack = []
    while len(stack) > 0:
        value = stack.pop()
        insertInTempStack(stack, value, tempStack)

    while len(tempStack) > 0:
        value = tempStack.pop()
        print value

def insertInTempStack(originalStack, value, tempStack):
    count = 0
    if len(tempStack) == 0:
        tempStack.append(value)
        return

    while 1 > 0:
        lenStack = len(tempStack)
        if lenStack == 0:
            break
        if tempStack[lenStack-1] < value:
            originalStack.append(tempStack[lenStack-1])
            tempStack.remove(tempStack[lenStack-1])
            count = count + 1
        else:
            break
    tempStack.append(value)

    while count > 0:
        tempStack.append(originalStack.pop())
        count = count - 1

stack = [34, 3, 31, 98, 92, 23]
sort(stack)

File: TopCoder/src/com/geeksforgeeks/stack/SortStackUsingTemporaryStack.java
package com.geeksforgeeks.stack;

import java.util.Stack;

/**
 * Created by sshil on 7/3/17.
 */
//http://www.geeksforgeeks.org/sort-stack-using-temporary-stack/
public class SortStackUsingTemporaryStack {

    public void sort(Stack<Integer> stack){
        Stack<Integer> tempStack = new Stack<>();
        while (!stack.isEmpty()){
            Integer current = stack.pop();
            insertInTemporaryStack(stack, current, tempStack);
        }
        while (!tempStack.isEmpty()){
            System.out.println(tempStack.pop());
        }
    }

    private void insertInTemporaryStack(Stack<Integer> original,
                                        Integer current,
                                        Stack<Integer> tempStack) {
        int count = 0;
        if (tempStack.isEmpty()){
            tempStack.push(current);
            return;
        }
        while (!tempStack.isEmpty() && tempStack.peek() < current){
            original.push(tempStack.pop());
            count++;
        }
        tempStack.push(current);
        for ( int i = 0 ; i  < count ; i++ ) {
            tempStack.push(original.pop());
        }
    }


    public static void main(String[] args) {
        Stack<Integer> stack = new Stack<>();
        stack.push(23);
        stack.push(92);
        stack.push(98);
        stack.push(31);
        stack.push(3);
        stack.push(34);
        new SortStackUsingTemporaryStack().sort(stack);


    }
}
